Update design, add editorial summary, search functionality and homepage

forabi opened the pull request:
* [x] Decide which animation library to use (if any)
* [x] Add `photoUrl` to Algolia index
* [x] Add skeleton component for search results while loading
* [ ] ~Add placeholder for search results page when box is empty~
* [x] Add placeholder for search results page when no results are found
* [x] Add skeleton component for Notable Person page
* [x] Add related people section
  * [x] Add related people to database
  * [x] Add related people to `NotablePerson` type
* [x] Add image attribution link
  * [x] Add image table to database
  * [x] Expose image information via API query
* [x] Make sources collapsable, collapse by default
* [x] Include Algolia logo in search page
* [x] Add home page
  * [x] Search box
  * [x] Add footer
* [x] Add polyfill for `IntersectionObserver`

Closes #221

wholesomedev commented on tslint.json:
```diff
@@ -1,5 +1,8 @@
-  ]
+  ],
+  "rules": {
+    "no-backbone-get-set-outside-model": false
```
lol

wholesomedev commented:
@forabi 

Here are a few comments on the UI side so far. I've classified the comments as "Future enhancements" (things that we can address later) and "Priority" (things we should prioritize).

# Future enhancements

* When I click on the back button on desktop, a gray border appears around the back button. That shouldn't happen
* Sometimes image of NP shows the body but not the face of the NP (example: https://hollowverse.com/Jack_Black)
* Nav bar scrolls until out of view then becomes sticky and switches to search bar
* Search text is too small (mismatch with design)
* Input field gets highlighted when I click on the magnifying glass
* Cover image doesn't look good on wider screens. It looks as a repeating pattern. I removed the blurring to check and I saw Steve Austin's repeating dome as a cover image: ![image](https://user-images.githubusercontent.com/26730388/34427774-b76c22bc-ebfa-11e7-9471-9d7fa36dc7f6.png)
* Quotation symbol is too strong, it clutters up the page
* "Other interesting profiles" section requires side scrolling on narrower displays. Is that what we want?
* Cards don't match design. These cards are elevated and have rounded corners. I think the flat cards are cleaner (and are closer to Facebook's design)
* Summary text is too crammed (needs bigger line-height or letter-spacing perhaps?)

# Priority

* Ryan Dunn page is empty
* Michael Jackson's image is missing
* Some times the Facebook Comments component is stuck on "Loading" (Steps to reproduce: load homepage, search for "st", select Steve Austin, scroll to bottom quickly and tap on one of the other interesting profiles, then scroll to bottom quickly and tap on another interesting profile. I did that a few times and Facebook comments wasn't loading any more) 
* Sources is always open
* On https://hollowverse.com/Chris_Martin page, at the bottom there's a link with the text "his wife". There are missing spaces around that text.

forabi commented:
> When I click on the back button on desktop, a gray border appears around the back button. That shouldn't happen

![screenshot_20171229_141459](https://user-images.githubusercontent.com/1012761/34436925-cff4b95e-eca2-11e7-9cf5-a83d940a99c5.png)

Are you referring to this? This is the default "focus ring" style applied by the browser for accessibility and keyboard navigation.

> Sometimes image of NP shows the body but not the face of the NP (example: https://hollowverse.com/Jack_Black)

We should use `smartcrop.js` to crop the images to 1:1 with face recognition enabled.

> Nav bar scrolls until out of view then becomes sticky and switches to search bar

What should happen instead?

> Input field gets highlighted when I click on the magnifying glass

I'm not sure what you mean. Is that happening on search page? Is the search input empty? If so, this is the intended behavior (input field is required and when empty, the browser highlights it and shows "Please fill out this field" for me (Chrome). Anyway, there are some issues with the navbar and search box. I'll need to do some refactoring here.

> Cover image doesn't look good on wider screens. It looks as a repeating pattern. I removed the blurring to check and I saw Steve Austin's repeating dome as a cover image: 
I'm going to prevent the default `background-repeat` behavior and make the image stretch (although I'm not sure it will look better). Any suggestions?

> Quotation symbol is too strong, it clutters up the page
I'm going to make the text color for the symbol lighter.

> "Other interesting profiles" section requires side scrolling on narrower displays. Is that what we want?
I thought you mentioned that was OK. I don't have a preference. We can shrink the width of each image so that it does not overflow.

> Cards don't match design. These cards are elevated and have rounded corners. I think the flat cards are cleaner (and are closer to Facebook's design)
Yeah, I started with the same styling as the design but noticed that it looks somewhat out of place when compared to the FB comments card and the editorial summary card (both have shadows, unlike the related people images):

![screenshot_20171225_213437](https://user-images.githubusercontent.com/1012761/34437099-1b19dfc6-eca4-11e7-9180-be93f971fa73.png)


Also, this is what Facebook suggestions look on mobile for me (notice the shadows and the wider gap):

![screenshot_20171225_224324](https://user-images.githubusercontent.com/1012761/34437095-0b940630-eca4-11e7-8a89-42175e4cc53f.png)

Anyway, it's your call really.

> Summary text is too crammed (needs bigger line-height or letter-spacing perhaps?)
I'll fix that.

## Priority

> Ryan Dunn page is empty

It's a stub on the old Hollowverse website. What should we do?

> Michael Jackson's image is missing
It turns out that because the image was manually replaced, the import script didn't know how to attach the source (and we are currently requiring that each image has a `sourceUrl` for attribution). I think I'll end up implementing some kind of override support for images so that we can force the scraper to use a specific image instead of searching for one using Wikipedia's API.

> Sometimes the Facebook Comments component is stuck on "Loading" (Steps to reproduce: load homepage, search for "st", select Steve Austin, scroll to bottom quickly and tap on one of the other interesting profiles, then scroll to bottom quickly and tap on another interesting profile. I did that a few times and Facebook comments wasn't loading any more)

I noticed that. I'll delay the call to `MutationObserver.takeRecords` until the promise of parsing the comments resolves and see if that fixes the issue.

> Sources is always open
This is already in the PR tasks. I'm working on it.

> On https://hollowverse.com/Chris_Martin page, at the bottom there's a link with the text "his wife". There are missing spaces around that text.

I thought https://github.com/hollowverse/scraper/pull/26 fixed this. I'm pretty sure I've removed all calls to `trim()` in the scrape function. I'll look into that again.

wholesomedev commented:
>![screenshot_20171229_141459](https://user-images.githubusercontent.com/1012761/34436925-cff4b95e-eca2-11e7-9cf5-a83d940a99c5.png)
>
>Are you referring to this? This is the default "focus ring" style applied by the browser for accessibility and keyboard navigation.

I don't see a similar pattern on Facebook or other websites. Can we eliminate this behavior?

>> Nav bar scrolls until out of view then becomes sticky and switches to search bar
>
>What should happen instead?

The nav bar should always be fully visible on the screen. And should animate into search bar when the user starts scrolling. That's how it is in `/design`

>> Input field gets highlighted when I click on the magnifying glass
>
>I'm not sure what you mean. Is that happening on search page? Is the search input empty? If so, this is the intended behavior (input field is required and when empty, the browser highlights it and shows "Please fill out this field" for me using Chrome). Anyway, there are some issues with the navbar and search box. I'll need to do some refactoring here.

I mean the gray border here 
![image](https://user-images.githubusercontent.com/26730388/34442558-bf08df88-ec77-11e7-8eee-b1dbb30a42d1.png)

>> Cover image doesn't look good on wider screens. It looks as a repeating pattern. I removed the blurring to check and I saw Steve Austin's repeating dome as a cover image: 
>
>I'm going to prevent the default `background-repeat` behavior and make the image stretch (although I'm not sure it will look better). Any suggestions?

I don't really have any suggestions. I'd say see if you can make something that you can be happy with. If not, I can also try to give it a shot.

>> "Other interesting profiles" section requires side scrolling on narrower displays. Is that what we want?
>
>I thought you mentioned that was OK. I don't have a preference. We can shrink the width of each image so that it does not overflow.

I think it's okay to keep it this way.

One thing I just noticed on my iPhone though is that when you swipe right or left, there's no inertial scrolling. The tiles make an abrupt stop instead. We need to add...

```
-webkit-overflow-scrolling: touch;
```

...to the scrolling element to enable inertial scrolling on those tiles.

>> Cards don't match design. These cards are elevated and have rounded corners. I think the flat cards are cleaner (and are closer to Facebook's design)
>
>Yeah, I started with the same styling as the design but noticed that it looks somewhat out of place when compared to the FB comments card and the editorial summary card (both have shadows, unlike the related people images):

Actually, my bad on this one. I just checked the design again and what you have matches it. I don't like the elevation though (generally speaking, for all cards), but that's something we can revisit later on.

>> Ryan Dunn page is empty
>
>It's a stub on the old Hollowverse website. What should we do?

Can we show some text on these types of pages? Something like

> Share what you know about the religion and political views of Ryan Dunn in the comments below

And can we prevent them from showing up in the search results?

forabi commented:
> I just checked the design again and what you have matches it.

I think you confused the screenshot in my last comment with what I ended up doing. The screenshot above is before I made the changes you currently see on the deployed branch. The deployed version does not really match what we have in `hollowverse/design`.

wholesomedev commented:
@forabi Hmmm, the cards in the deployed version look very similar to `hollowverse/design` from what I can tell. The only difference that I can see is a teeny tiny bit of extra elevation on the deployed version?

forabi commented:
> Hmmm, the cards in the deployed version look very similar to hollowverse/design from what I can tell. The only difference that I can see is a teeny tiny bit of extra elevation on the deployed version?

This and the gap between the image boxes is larger in the deployed version.

wholesomedev commented:
@forabi I see. Either design is fine by me for now.

forabi commented:
Tracking the progress of the issues so far:


# Future enhancements

* [x] When I click on the back button on desktop, a gray border appears around the back button. That shouldn't happen (removed focus outline)
* [ ] Sometimes image of NP shows the body but not the face of the NP (example: https://hollowverse.com/Jack_Black)
* [x] Nav bar scrolls until out of view then becomes sticky and switches to search bar (fixed)
* [ ] Search text is too small (mismatch with design) (I checked the design and the computed font size for the search box is 14px, same as the this PR)
* [x] Input field gets highlighted when I click on the magnifying glass (removed focus outline)
* [x] Cover image doesn't look good on wider screens. It looks as a repeating pattern. (made the image stretch for now)
* [x] Quotation symbol is too strong, it clutters up the page (tuned down color and font size)
* [x] "Other interesting profiles" section requires side scrolling on narrower displays. Is that what we want?
* [x] Cards don't match design. These cards are elevated and have rounded corners. I think the flat cards are cleaner (and are closer to Facebook's design)
* [x] Summary text is too crammed (needs bigger line-height or letter-spacing perhaps?) (Increase line height to match editorial summary)

# Priority

* [x] Ryan Dunn page is empty (Added pitch text when no editorial summary exists for NP)
* [ ] Michael Jackson's image is missing
* [ ] Sometimes the Facebook Comments component is stuck on "Loading" (fix attempted)
* [x] Sources is always open (fixed)
* [x] On https://hollowverse.com/Chris_Martin page, at the bottom there's a link with the text "his wife". There are missing spaces around that text. (Partially fixed, I've tracked down the issue to a bug in TypeORM query builder https://github.com/typeorm/typeorm/issues/1397). In the meantime, I've added extra spaces around each text node even if not needed, double white spaces are collapsed automatically (default browser behavior).

forabi commented:
@wholesomedev I've yet to find the cause of Facebook comments component being stuck at "Loading...". Other than that and the polyfill for `IntersectionObserver`, I think this is ready for review.

forabi commented:
The deployed version on the `new-app` branch has a potential fix for the Facebook comments component issue. Can you confirm that the issue is fixed?

wholesomedev commented on src/app/components/TextPage/TextPage.tsx:
```diff
@@ -0,0 +1,11 @@
+  className,
+  ...rest
+}: React.HTMLAttributes<HTMLDivElement>) => (
+  <div className={cc([className, classes.root])} {...rest} />
```
I like how you're passing `children` here. It's never occurred to me to pass `children` as a prop like this 👍

wholesomedev commented on src/app/hocs/AsyncComponent/AsyncComponent.tsx:
```diff
@@ -1,38 +1,65 @@
+   * as soon as the component mounts. This is only intended to avoid flashing
+   * of loading component in image components when using server side rendering
+   * and the image is already cached client-side.
+   * In such cases, the loading component would should for a brief moment before
```
typo: "should" => "flash"

But you don't mention how adding this delay prevents the flashing? Or does it only delay the flashing?

forabi replied:
I've expanded the comment to explain that in more details. Is it more clear now?

> If the delay is set to, say, 200ms, the browser has 200ms to show the image before we trigger the "loading" state. If the image is already in the browser cache, the browser will usually finish loading the image before the 200ms deadline, and the loading state will never trigger, thus we avoid flashing.

wholesomedev commented on src/app/hocs/AsyncComponent/AsyncComponent.tsx:
```diff
@@ -41,47 +68,61 @@ type State<T> = {
+        promises.push(
+          loadPromise.then(value => ({
+            value,
+            isInProgress: false as false,
```
What does `false as false` accomplish here?

forabi replied:
TypeScript is looking at this object and trying to infer its type, its considers `isInProgress` to be a `boolean`. It then tries to check if that type is compatible with any of the subtypes of `AsyncResult`. However, since `AsyncResult` types are very strict regarding the value of `isInProgress`, TypeScript complains that `isInProgress` as a `boolean` does not match any of the subtypes of `AsyncResult`.

forabi replied:
Anyway, it might be better to use a type assertion (`as Partial<SuccessResult<T>>`) instead.

wholesomedev commented on src/app/hocs/WithData/WithData.tsx:
```diff
@@ -0,0 +1,142 @@
+    result: getResolvedDataForKey(state)(ownProps.dataKey),
+  }),
+  { requestData },
+)(Wrapper);
```
👍

wholesomedev commented on src/app/pages/About/About.tsx:
```diff
@@ -0,0 +1,19 @@
+    <p>
+      Email me at <a href="mailto:msafi@msafi.com">msafi@msafi.com</a> if you’d
+      like to discuss anything.
+    </p>
```
Can this paragraph...

```jsx
<p>
      Email me at <a href="mailto:msafi@msafi.com">msafi@msafi.com</a> if you’d
      like to discuss anything.
</p>
```

...be changed to...

```jsx
<p>
      Email us at <a href="mailto:hello@hollowverse.com">hello@hollowverse.com</a> if you’d
      like to discuss anything.
</p>
```

forabi replied:
Done

wholesomedev commented on src/app/pages/NotablePerson/NotablePerson.tsx:
```diff
@@ -1,210 +1,171 @@
+                  )}
+                </article>
+                {notablePerson.relatedPeople.length ? (
+                  <div>
```
This component is growing too big. Should the `relatedPeople` functionality be spun off into its own component?

forabi replied:
Done.

wholesomedev commented on src/app/pages/NotablePerson/NotablePersonSkeleton.module.scss:
```diff
@@ -0,0 +1,111 @@
+  100% {
+    transform: translateX(780px);
+  }
+}
```
This approach of making skeleton components requires not a small amount of effort. The CSS of these components has to very closely match that of the fleshed component. Otherwise, skeleton components don't work as well.

I have an idea about creating skeleton components that might make this both easier and work better. I haven't put this idea to the test yet, but I think it might work.

What if instead of rendering a skeleton component while you wait for data, you render your original component but with _dummy data_ and _additional CSS_ to format the dummy data in a certain way to indicate that it's progressing, like render blurry text with animated gradient or a blurry photo with animated gradient. Regardless of how you style the dummy data, don't you think this approach would work?

forabi replied:
I thought about this before going with this. The problem is, sometimes we need to show the skeleton component in place of the actual component while that component is loading, like the NotablePerson component above. We cannot use the component while it's loading obviously. So the approach you suggested only works when we have the component loaded but we are loading the data it needs to display.

forabi replied:
By the way, the approach you described is exactly what https://github.com/lightsinthesky/react-skeleton does

wholesomedev replied:
Ah, interesting. Well, it doesn't have to be part of this PR but it's something we can continue to think about.

>The problem is that sometimes we need to show the skeleton component in place of the actual component while that component is loading, like the `NotablePerson` component above.

One way to solve this problem is to create `NotablePersonBody` component that can be rendered differently inside the `NotablePerson` component depending on the availability of data.

wholesomedev commented on src/app/pages/SearchResults/SearchResults.tsx:
```diff
@@ -0,0 +1,164 @@
+                }
+
+                return (
+                  <div>
```
I think breaking up this large render function either into multiple components or separate render methods would improve readability and comprehension here.

forabi replied:
Done

forabi commented on package.json:
```diff
@@ -5,31 +5,59 @@
+    "clean": "del-cli './src/app/**/*.module.scss.d.ts' './dist' && mkdir -p ./dist",
     "validate-filenames": "validate-filenames -c commonconfig.js",
     "lint-staged": "lint-staged"
   },
```
You can do:

```
API_ENDPOINT=https://api.hollowverse.com/graphql yarn dev
```

We can add shortcuts but I'm running out of meaningful names for the commands 😅

forabi replied:
Regarding `docker-compose`, `docker-compose up` is all we have to remember. It calls `yarn dev` in the Docker container.

wholesomedev replied:
how about `yarn dev-prod-api`? I know it's verbose, but having the command in `package.json` helps with memory for someone like me who's not as intimately involved with the setup as you are.

forabi replied:
Do you think it makes sense to default to `API_ENDPOINT` being the production API instead of the local instance? If so, we need to document that CORS has to be disabled during development.

wholesomedev replied:
I've been using 

```
API_ENDPOINT=https://api.hollowverse.com/graphql yarn dev
```

in my review of this PR.

There's an issue where every time I navigate between NP pages, the app breaks due to some issue with CORS preflight check. Then I click on the "Reload" button and the page loads. I'm not sure why that's happening. But if we can fix that, then I think it's fine to default `yarn dev` to run against production API.

forabi replied:
That means that the browser extension you are using is not doing anything. When you reload the page, what you is the server-rendered markup of the page. It has all the data because the server can freely access the API. When you navigate to a new page, the call to the API now happens in the browser and it fails because of CORS restrictions. You may need to add `https://api.hollowverse.com/graphql` to your extension exception list and keep DevTools open while testing (looks like it only works with DevTools open, not sure).

wholesomedev replied:
Okay, so now I tried this extension instead and it works fine: https://chrome.google.com/webstore/detail/cors-toggle/jioikioepegflmdnbocfhgmpmopmjkim/related?hl=en

So I'm okay with making `yarn dev` work against prod API by default

forabi commented on package.json:
```diff
@@ -5,31 +5,59 @@
+      "path": "./dist/client/**/*.svg",
+      "maxSize": "30 kB"
+    }
+  ],
```
`bundlesize` does not currently support a separate configuration file (https://github.com/siddharthkp/bundlesize/issues/162)

forabi commented on src/app/components/Collapsable/Collapsable.tsx:
```diff
@@ -0,0 +1,65 @@
+          checked={isOpen}
+          aria-checked={Boolean(isOpen)}
+        />
+        <label className={classes.label} {...{ for: id }}>
```
The input, label, and children container all have to be siblings in order to hide the children with pure CSS using the sibling selector:

```scss
.input:not(:checked) ~ .children {
  display: none;
}
```

If we wrap the `input` in a the `label`, we can no longer target the children and we'll need to use JS for that. We'll also need to replicate some of the browser behavior for accessibility.

> If so, isn't for renamed to htmlFor in JSX

Correct. Unfortunately, Preact prefers standard attribute names, e.g. `class`, `for`. `preact-compat` should convert `className` and `htmlFor` to their Preact equivalents and indeed it does so. However, I noticed that `htmlFor` only works for the browser build. There seems to be an issue in `preact/server` where the `htmlFor` attribute is passed as-is because when I disable JS, the generated server markup has `htmlfor` instead of `for` and the collapsable component does not toggle when I click on the label.

The `{...{ for: id }}` tricks TypeScript into ignoring the "unknown" property `for`, since TS expects `htmlFor` instead.

> wouldn't it be more reliable to use a UUID method from `lodash` instead of `random`

Lodash does not have a `uuid` function as far as I know. It does have `uniqueId` but that's just returns `i + 1` every time it's called. I started with `uniqueId` but since it's also called on the server, the number may grow indefinitely until the Node process restarts.

wholesomedev replied:
I see. Should this be documented?

With regard to UUID, I've used [this](https://stackoverflow.com/a/8809472/604296) in the past. Or we can continue to use `random`. No objections from me.

forabi replied:
I've replaced `random` with the suggested UUID function. Actually, the previous code had a problem: `random()` always returned either `0` or `1` (looks like those are the default parameters of the function)

forabi commented on src/app/components/EditorialSummary/EditorialSummary.tsx:
```diff
@@ -0,0 +1,200 @@
+  references = new Map<Node, Source>();
+
+  state: State = {
+    shouldShowSources: __IS_SERVER__,
```
The server markup will have the source shown by default (`shouldShowSources: true`). When JS is executed, Preact calls `render` and sees that, in the browser build, `shouldShowSources` is now `false`, so it hides them.

forabi commented on src/app/components/LazyImage/LazyImage.tsx:
```diff
@@ -0,0 +1,39 @@
+
+          return (
+            <div>
+              <Image {...imageProps} />
```
This is for styling consistency with `<noscript><Image /></noscript>` so that we do not assume that `Image` is a direct child.

forabi replied:
This has been updated since you reviewed it. It's no longer wrapped in a `div`.

forabi commented on src/app/components/NavBar/NavBar.tsx:
```diff
@@ -1,8 +1,134 @@
+  Props & RouteComponentProps<any>
+> {
+  goBack = (e: React.MouseEvent<HTMLElement>) => {
+    e.preventDefault();
```
Actually it's not required anymore. I was using a `Link` previously. I'm going to remove it.

forabi commented on src/app/components/NavBar/NavBar.tsx:
```diff
@@ -1,8 +1,134 @@
+                            to="/search"
+                          >
+                            <SvgIcon size={20} {...searchIcon} />
+                            <span className="sr-only">Search</span>
```
I'm following some of the basic accessibility recommendations provided by Lighthouse. TSLint also enforces some accessibility rules.

Generally speaking, we should at least check:

* Keyboard navigation
* Focusable elements have some visual indication
* Buttons, images and links are correctly labeled.
* Menus are correctly formatted (`ul` and `ol` have direct `li` children)
* Use of semantic markup. `nav`, `main`, `article`, headings... etc. This greatly helps screen readers.
* Use built-in browser components, like inputs and buttons, instead of replicating their behavior with JS/CSS. Most of the browser components have behaviors that are either hard or impossible to replicate with JS/CSS.
* Testing with screen readers/OS accessibility services.

Unfortunately Linux accessibility services are not great. I used to use ChromeVox as a screen reader for Chrome but it broke for some reason.

I tested some parts of the UI with Android's accessibility service and once I get my hands on an iPhone I'll do some further testing.

I've been following Rob Dodson's [a11ycasts](https://www.youtube.com/watch?v=V1yoOLhx_qA&list=PLNYkxOF6rcICWx0C9LVWWVqvHlYJyqw7g) for a while and it's great!.

forabi commented on src/app/components/NavBar/NavBarButton.module.scss:
```diff
@@ -0,0 +1,22 @@
+  border: 0 !important;
+  flex-grow: 0;
+  padding: 8px 15px !important;
+  background: transparent !important;
```
`NavBarButton` extends `Button`, these are need to override the default `Button` styles. I think there's room for improvement with regard to CSS.

forabi commented on src/app/components/NavBar/SearchView.tsx:
```diff
@@ -0,0 +1,102 @@
+  isSearchPage: boolean;
+  goToSearch(): any;
+  setShouldFocusSearch(isFocused: boolean): any;
+  searchQueryChanged({ query }: { query: string }): any;
```
Yeah, connecting `SearchView` is probably better.

forabi replied:
Done.

forabi commented on src/app/components/NavBar/SearchView.tsx:
```diff
@@ -0,0 +1,102 @@
+
+  focusIfNecessary = (props: Props = this.props) => {
+    if (props.isFocused && this.searchInput) {
+      this.searchInput.focus();
```
HTML has `autoFocus` on `input` but it has issues with React rendering.

forabi commented on src/app/components/PersonDetails/PersonDetails.module.scss:
```diff
@@ -1,51 +1,89 @@
+  }
+}
+
+@supports (filter: blur(5px)) {
```
`@supports` is native CSS, not Sass. Support is checked at runtime.

wholesomedev replied:
TIL!

forabi commented on src/app/components/Status/Status.tsx:
```diff
@@ -0,0 +1,57 @@
+      return null;
+    }
+  },
+);
```
These are used for SSR (https://github.com/forabi/hollowverse/blob/92729b6fbb24ae8a4f3c20d1b0b1e2b5fa70a7fc/src/app/server.tsx#L63). Usage is here https://github.com/forabi/hollowverse/blob/92729b6fbb24ae8a4f3c20d1b0b1e2b5fa70a7fc/src/app/pages/NotablePerson/NotablePerson.tsx#L83

wholesomedev replied:
Aha, the `Status` component renders nothing, and is only used for its side effect on the Redux store (which is then queried on the server-side). I was going to ask "so why don't you just use plain Redux actions", but I like how this component can be conveniently embedded within the JSX logic...

But...come to think of it more...you could also embed a function a function in the JSX logic...hmmm 🤔

forabi replied:
Yeah, but I've not seen many functions called in `render`. Actually this `Status` component is inspired by this https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/guides/server-rendering.md#404-401-or-any-other-status

forabi commented on src/app/components/PersonDetails/PersonDetails.tsx:
```diff
@@ -2,45 +2,64 @@ import * as React from 'react';
-      </div>
-    )}
+    <div className={classes.content}>
+      {photo ? (
```
I remember somebody on Twitter mentioned that `condition && <Component />` has a caveat. If `condition` is `true` and `<Component />` returns a falsey value, the whole expression evaluates to `true` and it might be coerced to a string (`"true"`) in some cases.

forabi commented on src/app/components/Post/Post.tsx:
```diff
@@ -0,0 +1,86 @@
+    text: string;
+  }>;
+  children?: React.ReactNode;
+};
```
I generally prefer to keep components completely unaware of the data layer. It makes them easier to test in the future. I like to think that the component props just _happen to_ match the API types.

forabi commented on src/app/components/NavBar/SearchView.tsx:
```diff
@@ -0,0 +1,102 @@
+              <span className="sr-only">Search</span>
+            </NavBarLink>
+          </Route>
+        </Switch>
```
The React Router guys have been pushing for the idea that `Route`s are just components and there's nothing magical about them. Actually, `Route` can be simply replaced with an `if` statement (provided that we can access the `location` prop somehow).

> All you really need here is to know the route where the user currently is, right?
I don't see how that could be done if not using React Router?

wholesomedev replied:
It's just a bit hard to reason about when encountered for the first time. I found myself staring at the code for a good while until I finally got what's going on. With that said, I'm actually okay with it (as long as it's not a non-standard practice). It's just that I'm unfamiliar with the pattern.

> provided that we can access the `location` prop somehow

I think `react-router-redux` puts the router state in Redux. That way you can connect the component to the store and get the `location` prop from there.

forabi replied:
> I think react-router-redux puts the router state in Redux. That way you can connect the component to the store and get the location prop from there.

I was just stating that we need the `location` prop for this. React Router export a HOC, `withRouter`, that provides `location`.

forabi commented on src/app/components/Collapsable/Collapsable.module.scss:
```diff
@@ -0,0 +1,51 @@
+      }
+    }
+  }
+}
```
The icon rotates when the `Collapsable` state changes. Any other suggestions? I feel we are pushing the limits of CSS animations and we might need to bring some animation library in the future. CSS animations has the benefit of working without JS though (and it usually performs better too).

wholesomedev replied:
I was going to suggest animating between `display: none` and `display: inline-flex` or animating between `height: 0` and `height: auto`, but a quick search revealed that that's not easily possible...so I'm not sure...

forabi replied:
Not to mention that animating [`height` will trigger layout every single frame](https://csstriggers.com/height). Unlike things like `opacity` and `transform`, `height` is not one of the cheap properties that can be offloaded to the GPU.

forabi commented on src/app/components/NavBar/NavBar.tsx:
```diff
@@ -1,8 +1,134 @@
+                <Switch>
+                  <Route path="/search">{searchView}</Route>
+                  <Route>
+                    {() => {
```
When using the function version, the JSX expressions are only evaluated when the route matches. So this could be better performance-wise. It's not the reason why I used an anonymous function though. I may have forgot to remove it during refactoring.

What do you think?

wholesomedev replied:
I think the anonymous function makes it seems as though there's something special going on here, which isn't the case.

forabi replied:
Removed the anonymous function

forabi commented on src/app/components/App/App.tsx:
```diff
@@ -1,24 +1,41 @@
+import { LoadableSearchResults } from 'pages/SearchResults/LoadableSearchResults';
+import { LoadableAbout } from 'pages/About/LoadableAbout';
+import { LoadablePrivacyPolicy } from 'pages/PrivacyPolicy/LoadablePrivacyPolicy';
+import { LoadableHome } from 'pages/Home/LoadableHome';
```
That would not work.

To get Webpack to code split and create a chunk for the loadable component, we need to use the _dynamic_  `import()` with the path to the component. If we import the component and pass it to a HOC to make it loadable... well, it's not going to be in another chunk, because we just imported it using the standard `import { Component } from 'path/to/Component'`.


It wouldn't work with a HOC that accepts the _path_ to the file either. The dynamic `import()` call must be statistically analyzable so that Webpack knows what files to include in the dynamic chunks. If we create a HOC for loadable components, this is what `import()` call could look like:

```typescript
import(props.pathToComponent)
```

Webpack will have no idea where to find the files at build time. It's actually pretty smart and _can_ create chunks if the string is partially static:

```typescript
import(`components/${componentName}`);
```

However, in that case, Webpack will go ahead and create chunks for every single file that is found in `components`. We can limit the files pattern using Webpack's [magic comments](https://medium.com/faceyspacey/how-to-use-webpacks-new-magic-comment-feature-with-react-universal-component-ssr-a38fd3e296a) but it feels a little bit hacky to me. I don't like writing build tool-specific comments in code.

forabi commented on src/app/components/Footer/Footer.tsx:
```diff
@@ -0,0 +1,35 @@
+        <Link to="/privacy-policy">Privacy</Link>
+      </li>
+      <li>
+        <a href="https://twitter.com/hollowverse">Twitter</a>
```
Removed.

forabi commented on src/app/assets/algolia/search-by-algolia.svg:
```diff
@@ -0,0 +1 @@
+<svg viewBox="0 0 130 18" xmlns="http://www.w3.org/2000/svg"><title>search-by-algolia</title><defs><linearGradient x1="-36.868%" y1="134.936%" x2="129.432%" y2="-27.7%" id="a"><stop stop-color="#00AEFF" offset="0%"/><stop stop-color="#3369E7" offset="100%"/></linearGradient></defs><g fill="none" fill-rule="evenodd"><path d="M59.399.022h13.299a2.372 2.372 0 0 1 2.377 2.364V15.62a2.372 2.372 0 0 1- ...
```
I just removed the unused ones.

forabi commented on src/app/components/EditorialSummary/EditorialSummary.module.scss:
```diff
@@ -0,0 +1,39 @@
+      background-color: $color-blue-opacity;
+    }
+  }
+}
```
Sounds good! I'll see where we can apply that.

forabi commented on src/app/components/LazyImage/LazyImage.tsx:
```diff
@@ -0,0 +1,39 @@
+    const { rootMargin, ...imageProps } = this.props;
+
+    return (
+      <OptionalIntersectionObserver rootMargin="25%" triggerOnce>
```
Interesting. It's weird that it's actually working. You can confirm by looking at the Network tab and see that the requests for the related people images start as soon as you scroll down to the collapsed sources label. Could `react-intersection-observer` be doing something to make that work by default?

forabi commented on src/app/components/NavBar/NavBar.tsx:
```diff
@@ -1,8 +1,134 @@
+export type DispatchProps = {
+  goToSearch(): any;
+  searchQueryChanged({ query }: { query: string }): any;
+  setShouldFocusSearch(isFocused: boolean): any;
```
See https://github.com/hollowverse/hollowverse/pull/225#discussion_r159154766

> I generally prefer to keep components completely unaware of the data layer. It makes them easier to test in the future. I like to think that the component props just happen to match the API types.

wholesomedev commented on src/app/components/EditorialSummary/EditorialSummary.tsx:
```diff
@@ -0,0 +1,200 @@
+                      role="button"
+                      aria-label="Go back to reference"
+                    >
+                      ↩
```
But there's probably some styling that we can apply that will prevent Mac from rendering an emoji like the above...I say that because in the code, I see plain arrow, only when it is rendered on the screen do I see the emoji arrow. That leads me to believe there's certain styling/font perhaps that keeps the plain arrow?

forabi replied:
This is relevent: https://stackoverflow.com/questions/32915485/how-to-prevent-unicode-characters-from-rendering-as-emoji-in-html-from-javascrip

> Update: It seems (from comments below) there is a text presentation selector, FE0E, available to enforce text-vs-emoji. The selector is concatenated as a suffix without space onto the character's code, such as &#x231b;&#xFE0E for HTML hex or \u231B\uFE0E for JS. However, it is simply not honored by all browsers (eg Chrome and Edge).

forabi commented on src/app/components/NavBar/NavBar.tsx:
```diff
@@ -1,8 +1,134 @@
-    <a href="/">{props.title}</a>
-  </div>
-);
+import { Sticky } from 'components/Sticky/Sticky';
```
Why do you think it's named? It _is_ responsible for applying the sticky styles depending on the intersection state changes.

forabi replied:
misnamed*

wholesomedev replied:
Okay, I find it hard to tell what `Sticky` is doing...

It's a `position: relative` element. Inside of it, there's a `position: fixed` element, which is _always sticky_. But the children of this, always sticky, element receive a flag called `isSticky`, which is either `true` or `false` based whether the `position: relative` container is in view?

It seems like the `Sticky` component is trying to relay information between two elements where one element is interested in the position of another element. So it's not really about stickiness...

forabi replied:
> there's a position: fixed element, which is always sticky

Correct. But you forgot about the placeholder element, which is actually responsible for maintaining the original element's space while the element is sticking. So what `Sticky` is doing is:

1. Make the target element `fixed`.
2. Create a placeholder element to maintain the space of the target element.
3. Now the two elements are laid in the same space. The target element is on top because `fixed` creates a new stacking context.
4. Now when the user scrolls, the placeholder will go out of view, while the target will stay `fixed` at the top of the viewport. We monitor the placeholder element so that when it's not in view, it means that the target element has become effectively sticky from the user's point of view.

wholesomedev commented on src/app/store/features/data/actions.ts:
```diff
@@ -0,0 +1,4 @@
+import { createActionCreator } from 'store/helpers';
```
I see how `features/data` is being used with the `WithData` component and it makes sense...but everything is `data`, so can we differentiate this with a more descriptive name? Or at least add documentation or a README.md file in this folder?

forabi replied:
I changed the name to `asyncData` and added a README.md to document how this feature works with `WithData`.

wholesomedev commented on src/app/store/features/search/updateUrlEpic.ts:
```diff
@@ -0,0 +1,50 @@
+        search: searchParams.toString(),
+      };
+
+      const location = store.getState().routing.location;
```
Does it matter if you use an existing selector here or not?

forabi replied:
Updated

wholesomedev commented on src/app/store/helpers.ts:
```diff
@@ -0,0 +1,103 @@
+ */
+export function createAction<T extends ActionType>(
+  type: T,
+  payload: TypeToPayload[T],
```
Man, naming things _is_ hard. This should really be called `ReduxActionTypeToPayloadTypeScriptType` but naming it that will probably cause a riot 😂

wholesomedev commented on src/app/store/helpers.ts:
```diff
@@ -0,0 +1,103 @@
+}
+
+/** A better typed version of `redux-action`'s `createAction`.
+ * It only accepts valid action types and returs an action creator
```
`returs` => `returns`

forabi replied:
Fixed

wholesomedev commented on src/app/store/helpers.ts:
```diff
@@ -0,0 +1,103 @@
+
+    return defaultState;
+  };
+}
```
👍

wholesomedev commented:
I have a few comments about UI, but I don't think these comments should block this PR. This PR has grown too big and I think we should merge it ASAP and continue work in smaller chunks:

* I think we need to increase the size of the search box text
* "Jennifer Love Hewitt" card is bigger than other NPs on Jerry Lewis's page
* Margin around FB comments is too big
* In `hollowverse/design` the cover image has default highlights on the side, we're missing that here (I think we still haven't gotten the cover image right. I think we need to work more on it)
* On iPhone, when user focuses an input field (e.g. search box or Facebook comments box), the page zooms in for no reason
* When CORS is disabled, and I click on the "Reload" page button, the page registers two browser history items when it loads, so I can't back out of it without pressing the back button twice
* Clicks outside of the NP image redirect to Wikipedia. I think only clicks inside the NP's image should redirect to Wikipedia
* Issues on Ashanti (singer) page:
  * The search icon doesn't show on a hard reload
  * The cover image doesn't show. If I navigate from Ashanti to Keyshia Cole's page and then back to Ashanti, I see Keyshia's cover image on Ashanti's page

forabi commented:
I fixed some of these issues. Let's merge this and continue iterating in future PRs.

* [x] I think we need to increase the size of the search box text
* [x] "Jennifer Love Hewitt" card is bigger than other NPs on Jerry Lewis's page
* [x] Margin around FB comments is too big
* [ ] In `hollowverse/design` the cover image has default highlights on the side, we're missing that here (I think we still haven't gotten the cover image right. I think we need to work more on it)
* [ ] On iPhone, when user focuses an input field (e.g. search box or Facebook comments box), the page zooms in for no reason
* [ ] When CORS is disabled, and I click on the "Reload" page button, the page registers two browser history items when it loads, so I can't back out of it without pressing the back button twice
* [ ] Clicks outside of the NP image redirect to Wikipedia. I think only clicks inside the NP's image should redirect to Wikipedia
* [ ] Issues on Ashanti (singer) page:
  * [ ] The search icon doesn't show on a hard reload
  * [x] The cover image doesn't show. If I navigate from Ashanti to Keyshia Cole's page and then back to Ashanti, I see Keyshia's cover image on Ashanti's page

forabi commented on src/app/store/features/analytics/epic.ts:
```diff
@@ -0,0 +1,38 @@
+
+const initScript = once(async () => {
+  await importGlobalScript(GA_URL);
+  ga('create', 'UA-110141722-1', 'auto');
```
TypeScript has a concept of "module augmentation". Basically it allows modules to extend the prototype of another module.

I assume RxJS imports are written this way:

```
// module 'rxjs/add/map'

import { Observable } from 'rxjs/Observable';

Observable.map = /** ... */

```

However, TypeScript has an issue where if you import one operator in File A, it will assume that every other file now can access that operator on `Observable`. That's not always correct because we could try to access the operator in File B before we import File A. It's not aware of the order of dependencies and has no idea that we are doing things like code splitting.

That's where `rxjs-tslint-rules` help. The rules enforce that every operator is imported in the same module it's used in and that there are no unused imports.

forabi commented:
I've changed the default `API_ENDPOINT` to the production API and updated the wiki to include a note on CORS issue.
